src/meta_ops.rs

```diff
@@ -396,3 +400,296 @@ pub fn equal<'gc>(
+                })?
+            }
+        }
+
+        (Value::Table(t), a) => {
+            let get_sub = |t: Table<'gc>| {
+                let eq = t
+                    .metatable()
+                    .map(|t| t.get(ctx, MetaMethod::Sub))
+                    .unwrap_or_default();
+                if eq.is_nil() {
+                    None
+                } else {
+                    Some(eq)
+                }
+            };
+            if let Some(a_eq) = get_sub(t) {
+                MetaResult::Call(MetaCall {
+                    function: call(ctx, a_eq)?,
+                    args: [t.into(), a.into()],
+                })
+            } else {
+                Err(TypeError {
+                    expected,
+                    found: "table with no __sub",
+                })?
+            }
+        }
+        (Value::UserData(t), a) => {
+            let get_sub = |t: UserData<'gc>| {
+                let eq = t
+                    .metatable()
+                    .map(|t| t.get(ctx, MetaMethod::Sub))
+                    .unwrap_or_default();
+                if eq.is_nil() {
+                    None
+                } else {
+                    Some(eq)
+                }
+            };
+            if let Some(a_eq) = get_sub(t) {
+                MetaResult::Call(MetaCall {
+                    function: call(ctx, a_eq)?,
+                    args: [t.into(), a.into()],
+                })
+            } else {
+                Err(TypeError {
+                    expected,
+                    found: "userdata with no __sub",
+                })?
+            }
+        }
+
+        (a, Value::Table(t)) => {
+            let get_sub = |t: Table<'gc>| {
+                let eq = t
+                    .metatable()
+                    .map(|t| t.get(ctx, MetaMethod::Sub))
+                    .unwrap_or_default();
+                if eq.is_nil() {
+                    None
+                } else {
+                    Some(eq)
+                }
+            };
+            if let Some(a_eq) = get_sub(t) {
+                MetaResult::Call(MetaCall {
+                    function: call(ctx, a_eq)?,
+                    args: [a.into(), t.into()],
+                })
+            } else {
+                Err(TypeError {
+                    expected,
+                    found: "table with no __sub",
+                })?
+            }
+        }
+        (a, Value::UserData(t)) => {
+            let get_sub = |t| get_metamethod_from_userdata(ctx, t, MetaMethod::Sub);
+            if let Some(a_eq) = get_sub(t) {
+                MetaResult::Call(MetaCall {
+                    function: call(ctx, a_eq)?,
+                    args: [a.into(), t.into()],
+                })
+            } else {
+                Err(TypeError {
+                    expected,
+                    found: "userdata with no __sub",
+                })?
+            }
+        }
+
+        (Value::Nil, _) => Err(TypeError {
```

Comment:
these could probably be merged into 1 case, that uses `Value::type_name` (iirc) for the found typename.